### PR TITLE
Finish implementation for the mass-import of Legacy packages

### PR DESCRIPTION
**Pull request by f-f**
This PR implements the last bits that we were missing in order to have a mass-importer of "legacy" packages into the new Registry.
I had to shuffle a few things around (e.g. adjust what's returned by the topological sort, or how we print SemVer versions), but now it's possible to run
```bash
spago run -m Registry.Scripts.LegacyImport
```
...and the script will:
1. read all the legacy packages in
2. upload all the versions of the core packages to the storage backend
3. write out the `Metadata` (with the correct hashes) and the `registry-index` files

(the files produced by the last step have to be committed manually, the auto-commit is only for the CI)

Right now we limit this to only work on core packages (until we sort out the license issues for the rest of the packages), but extending this to all packages is only a matter of removing the filtering in the `LegacyImport` module.

We could consider this a fix for #215, but I guess we can leave that open until we actually push all the packages there?

### CODE REVIEW

File: ci/src/Registry/PackageGraph.purs [hunk 1]
```diff
@@ -23,6 +23,12 @@ import Registry.PackageName (PackageName)
 import Registry.PackageName as PackageName
 import Registry.Schema (Manifest)
 
+type Node =
+  { manifest :: Manifest
+  , packageName :: PackageName
+  , version :: SemVer
+  }
```
Comment (thomashoneyman) [inline ci/src/Registry/PackageGraph.purs]: Do you mind explaining why we need to introduce this new `Node` type? After all, a `Manifest` already includes the package name and version with the types used in `Node`:

https://github.com/purescript/registry/blob/05fa80c2bd1b95b9f470f9db8f223ba89577865a/ci/src/Registry/Schema.purs#L17-L23

From what I can tell, we don't need for a `Node` and the included `Manifest` to ever have different package names or versions, so it seems like it's just duplicated information.

File: ci/src/Registry/Scripts/LegacyImport/Manifest.purs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      let
+        rewrite = case _ of
+          "Apache 2" -> "Apache-2.0"
+          "Apache-2" -> "Apache-2.0"
+          "Apache 2.0" -> "Apache-2.0"
+          "BSD" -> "BSD-3-Clause"
+          "BSD3" -> "BSD-3-Clause"
+          "BSD-3" -> "BSD-3-Clause"
+          "3-Clause BSD" -> "BSD-3-Clause"
+          other -> other
```
Comment (thomashoneyman) [inline ci/src/Registry/Scripts/LegacyImport/Manifest.purs:45]: Unrelated to this PR, but I think we should disable any modification of package information when uploading packages, at least until we have the legal side of things confirmed.

**Comment by thomashoneyman** [comment]
Also unrelated to this PR: are the hashes we're using compatible with what Nix requires? I can't remember if Nix has a specific hash format. If so, it would be _really_ nice to be compatible with Nix out of the box.

**Review by thomashoneyman — CHANGES_REQUESTED** [review]
Everything looks good except for the introduction of the `Node` type. Once we chat about whether & why that's necessary to add, or revert it if it isn't necessary, then 👍 from me.

File: ci/src/Registry/Scripts/LegacyImport/Manifest.purs (hunk 2 above)
Comment (f-f) [reply ci/src/Registry/Scripts/LegacyImport/Manifest.purs:45]: We are limiting the upload to core packages for now, so we probably don't really have to worry about this yet:
https://github.com/purescript/registry/blob/e55de07f6f5c91390d77d0d4d07edad64b7d6fcb/ci/src/Registry/Scripts/LegacyImport.purs#L64-L70

File: ci/src/Registry/PackageGraph.purs (hunk 1 above)
Comment (f-f) [reply ci/src/Registry/PackageGraph.purs]: I am actually not sure why I went on this trip of adding this type - you're absolutely right, it's useless and I removed it :smile:

**Comment by f-f** [comment]
Thanks for the review @thomashoneyman! Everything should be addressed now.

> Also unrelated to this PR: are the hashes we're using compatible with what Nix requires?

When implementing the hashing I made sure that our code computes the same result as the `sha256sum` command (but don't trust me on this, we actually need to add some tests :smile:).
It looks like that's compatible with nix hashes:
```bash
$ nix-hash --flat --type sha256 package.json
d6d115122eba9d81410cd6cd7b88205f8e6515ac88d30aa29a5bb17ef4908580
$ sha256sum package.json
d6d115122eba9d81410cd6cd7b88205f8e6515ac88d30aa29a5bb17ef4908580  package.json
```

**Review by thomashoneyman — APPROVED** [review]

**Pull request merged**